internal/civisibility/integrations/gotesting/instrumentation.go

```diff
@@ -128,492 +128,344 @@ func deleteTestMetadata(tb testing.TB) {
-	if !isCiVisibilityEnabled() {
-		return
-	}
+						fmt.Printf("    [ %v after %v retries by Datadog's auto test retries ]\n", status, executionIndex)
+					}
 
-	// Get the CI Visibility span and check if we can mark it as skipped and close it
-	ciTestItem := getTestMetadata(tb)
-	if ciTestItem != nil && ciTestItem.test != nil && ciTestItem.skipped.CompareAndSwap(0, 1) {
-		ciTestItem.test.CloseWithFinishTimeAndSkipReason(integrations.ResultStatusSkip, time.Now(), skipReason)
+					// Check if total retry count was exceeded
+					if flakyRetrySettings.RemainingTotalRetryCount < 1 {
+						fmt.Println("    the maximum number of total retries was exceeded.")
+					}
+				},
+				// function to modify the execution metadata for each execution
+				nil, // No execMetaAdjust needed
+			)
+		}
 	}
+	return targetFunc
 }
 
-// instrumentSkipNow helper function to close and skip a `*testing.T, *testing.B, *testing.common` CI Visibility span
-func instrumentSkipNow(tb testing.TB) {
-	// Check if CI Visibility was disabled using the kill switch before
-	if !isCiVisibilityEnabled() {
-		return
-	}
+// applyEarlyFlakeDetectionAdditionalFeature applies the early flake detection feature as a wrapper of a func(*testing.T)
+func applyEarlyFlakeDetectionAdditionalFeature(testInfo *commonInfo, targetFunc func(*testing.T), settings *net.SettingsResponseData) func(*testing.T) {
+	earlyFlakeDetectionData := integrations.GetEarlyFlakeDetectionSettings()
+	if earlyFlakeDetectionData != nil &&
+		len(earlyFlakeDetectionData.Tests) > 0 {
 
-	// Get the CI Visibility span and check if we can mark it as skipped and close it
-	ciTestItem := getTestMetadata(tb)
-	if ciTestItem != nil && ciTestItem.test != nil && ciTestItem.skipped.CompareAndSwap(0, 1) {
-		ciTestItem.test.Close(integrations.ResultStatusSkip)
-	}
-}
+		// Define is a known test flag
+		isAKnownTest := false
 
-// instrumentTestingBFunc helper function to instrument a benchmark function func(*testing.B)
-func instrumentTestingBFunc(pb *testing.B, name string, f func(*testing.B)) (string, func(*testing.B)) {
-	// Check if CI Visibility was disabled using the kill switch before instrumenting
-	if !isCiVisibilityEnabled() {
-		return name, f
-	}
+		// Check if the test is a known test or a new one
+		if knownSuites, ok := earlyFlakeDetectionData.Tests[testInfo.moduleName]; ok {
+			if knownTests, ok := knownSuites[testInfo.suiteName]; ok {
+				if slices.Contains(knownTests, testInfo.testName) {
+					isAKnownTest = true
+				}
+			}
+		}
 
-	// Reflect the function to obtain its pointer.
-	fReflect := reflect.Indirect(reflect.ValueOf(f))
-	moduleName, suiteName := utils.GetModuleAndSuiteName(fReflect.Pointer())
-	originalFunc := runtime.FuncForPC(fReflect.Pointer())
+		// If it's a new test, then we apply the EFD wrapper
+		if !isAKnownTest {
+			return func(t *testing.T) {
+				var testPassCount, testSkipCount, testFailCount int
+
+				runTestWithRetry(
```

Comment:
Just a suggestion: this function takes a lot of arguments and many of them are whole functions, which makes it a bit hard to read since Go doesn't have keyword arguments. Maybe it would make sense to create a struct type to hold these arguments, so you could write something like:
```
testWithRetry{
    targetFunc: targetFunc,
    test: t,
    initialRetryCount: 0,
    ...
}.Run()
````

Reply:
good idea, I did the change in [1077d6e](https://github.com/DataDog/dd-trace-go/pull/2916/commits/1077d6e72f23d4d1715d36ffa7aef62984d046b3)